src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp

```diff
@@ -4553,6 +4607,10 @@ CTranslatorDXLToPlStmt::TranslateDXLDynForeignScan(
 	ForeignScan *foreign_scan_first_part =
 		gpdb::CreateForeignScan(oid_first_child, index, qual, targetlist,
 								m_dxl_to_plstmt_context->m_orig_query, rte);
+	const IMDRelation *md_rel_first_part =
+		m_md_accessor->RetrieveRel((*parts)[0]);
+	CheckForeignScanDistributionMismatch(foreign_scan_first_part,
```

Comment:
What about creating a function which makes it impossible to create ForeignScan without check?